File: src/main/java/de/tectoast/emolga/utils/sql/base/columns/SQLColumn.java
package de.tectoast.emolga.utils.sql.base.columns;

import de.tectoast.emolga.utils.sql.base.DataManager;

import java.sql.ResultSet;

public abstract class SQLColumn<T> {
    public final String name;
    public final DataManager manager;

    public SQLColumn(String name, DataManager manager) {
        this.name = name;
        this.manager = manager;
    }

    public String check(Object value) {
        return name + " = " + wrap(value);
    }

    public String wrap(Object value) {
        if (value == null) {
            return "NULL";
        }
        return value.toString();
    }

    public T retrieveValue(SQLColumn<?> checkcolumn, Object checkvalue) {
        return manager.read(manager.select(checkcolumn.check(checkvalue), this), rs -> {
            if (rs.next()) {
                return getValue(rs);
            }
            return null;
        });
    }

    public ResultSet getAll(T value) {
        return manager.read(manager.selectAll(check(value)), rs -> rs);
    }

    public ResultSet getSingle(T value) {
        return manager.read(manager.selectAll(check(value)), rs -> {
            rs.next();
            return rs;
        });
    }

    public boolean isAny(Object value) {
        return retrieveValue(this, value) != null;
    }


    public abstract T getValue(ResultSet set);

    public abstract void update(ResultSet set, Object value);
}

File: src/main/kotlin/de/tectoast/emolga/utils/sql/base/columns/SQLColumn.kt
package de.tectoast.emolga.utils.sql.base.columns

import de.tectoast.emolga.utils.sql.base.DataManager
import java.sql.ResultSet

abstract class SQLColumn<T>(val name: String, val manager: DataManager) {
    fun check(value: Any?): String {
        return name + " = " + wrap(value)
    }

    open fun wrap(value: Any?): String {
        return value?.toString() ?: "NULL"
    }

    open fun retrieveValue(checkcolumn: SQLColumn<*>, checkvalue: Any): T? {
        return DataManager.read<T?>(manager.select(checkcolumn.check(checkvalue), this)) { rs: ResultSet ->
            if (rs.next())
                return@read getValue(rs)
            else
                null
        }
    }

    fun getAll(value: T): ResultSet? {
        return DataManager.read<ResultSet?>(manager.selectAll(check(value))) { rs: ResultSet? -> rs }
    }

    fun isAny(value: Any): Boolean {
        return retrieveValue(this, value) != null
    }

    abstract fun getValue(set: ResultSet): T
}